File: HoYa/Entities/WorkPlan.cs
using Newtonsoft.Json;
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace HoYa.Entities
{
    public class WorkPlan : Instance
    {
        public virtual Guid? ItemId { get; set; }
        [ForeignKey("ItemId")]
        public virtual Item Item { get; set; }
        public virtual Guid? WorkPlanRecordId { get; set; }
        [ForeignKey("WorkPlanRecordId")]
        public virtual WorkPlanRecord WorkPlanRecord { get; set; }


        public DateTime? StartDate { get; set; }//後端填寫

        public virtual Guid? StartId { get; set; }//後端填寫
        [ForeignKey("StartId")]
        [JsonIgnore]
        public virtual WorkEvent Start { get; set; }//後端填寫


        public DateTime? EndDate { get; set; }//後端填寫

        public virtual Guid? EndId { get; set; }//後端填寫
        [ForeignKey("EndId")]
        [JsonIgnore]
        public virtual WorkEvent End { get; set; }//後端填寫


        public WorkPlan()
        {
        } 
    }

    public class WorkPlanRecord : Detail<WorkPlan>
    {
        public decimal? Value { get; set; }
        public virtual Guid? UnitId { get; set; }
        [ForeignKey("UnitId")]
        public virtual Inventory Unit { get; set; }
        public int Sort { get; set; }
        public WorkPlanRecord()
        {
        }
    }

    public class WorkOrder : Detail<WorkPlan>
    {
        public virtual Guid? ItemId { get; set; }
        [ForeignKey("ItemId")]
        public virtual Item Item { get; set; }

        public DateTime? StartDate { get; set; }//後端填寫

        public virtual Guid? StartId { get; set; }//後端填寫
        [ForeignKey("StartId")]
        [JsonIgnore]
        public virtual WorkEvent Start { get; set; }//後端填寫


        public DateTime? EndDate { get; set; }//後端填寫

        public virtual Guid? EndId { get; set; }//後端填寫
        [ForeignKey("EndId")]
        [JsonIgnore]
        public virtual WorkEvent End { get; set; }//後端填寫  

        public bool? Archived { get; set; }
        public decimal? Value { get; set; }
        public int Sort { get; set; } 
        public WorkOrder(string ownerId)
        {
            this.OwnerId = Guid.Parse(ownerId);
        }
    }

    public class WorkEvent : Detail<WorkOrder>
    {
        public int Sort { get; set; }

        public virtual Guid? TypeId { get; set; }
        [ForeignKey("TypeId")]
        public virtual Option Type { get; set; }

        public WorkEvent(string ownerId, string typeId)
        {
            this.OwnerId = Guid.Parse(ownerId);
            this.TypeId = Guid.Parse(typeId);
        } 
    }

    public class WorkEventInventory : Relation<WorkEvent,Inventory>//Position在Station上
    {
        public int Sort { get; set; }

        public virtual Guid? TypeId { get; set; }//Input(Position=Station), Output(Position=Station), Pickup(Position=Person)
        [ForeignKey("TypeId")]
        public virtual Option Type { get; set; }

        public WorkEventInventory(string ownerId, string targetId)
        {
            this.OwnerId = Guid.Parse(ownerId);
            this.TargetId = Guid.Parse(targetId);
        }
    }
     

    public class StationRecord : Relation<Inventory,WorkEvent>
    {
        public int Sort { get; set; }
        public StationRecord(string ownerId, string targetId)
        {
            this.OwnerId = Guid.Parse(ownerId);
            this.TargetId = Guid.Parse(targetId);
        }
    } 
}

File: HoYa.Web/src/@entities/workPlan.ts
import { Instance, Detail, Option, Relation } from "./entity";
import { Item } from "./item";
import { Inventory } from "./inventory";

export class WorkPlan extends Instance {
    itemId: string;
    item: Item;
    workPlanRecordId: string;
    workPlanRecord: WorkPlanRecord;
    startDate: string;//後端填寫
    startId: string;//後端填寫
    start: WorkEvent;//後端填寫
    endDate: string;//後端填寫
    endId: string;//後端填寫
    end: WorkEvent;//後端填寫

    constructor() {
        super();
    }
}

export class WorkPlanRecord extends Detail<WorkPlan>
{
    value: number;
    sort: number;

    constructor() {
        super();
    }
}

export class WorkOrder extends Detail<WorkPlan>
{
    itemId: string;
    item: Item;
    startDate: string;//後端填寫
    startId: string;//後端填寫
    start: WorkEvent;//後端填寫
    endDate: string;//後端填寫
    endId: string;//後端填寫
    end: WorkEvent;//後端填寫  
    archived: boolean;
    value: string;
    sort: number;

    constructor(ownerId: string) {
        super();
        this.ownerId = ownerId;
    }
}

export class WorkEvent extends Detail<WorkOrder>
{
    sort: string;
    typeId: string;
    type: Option;

    constructor(ownerId: string, typeId: string) {
        super();
        this.ownerId = ownerId;
        this.typeId = typeId;
    }
}

export class WorkEventInventory extends Relation<WorkEvent, Inventory>//Position在Station上
{
    sort: number;
    typeId: string;//Input(Position=Station), Output(Position=Station), Pickup(Position=Person)
    type: Option;

    constructor(ownerId: string, typeId: string) {
        super();
        this.ownerId = ownerId;
        this.typeId = typeId;
    }
}


export class StationRecord extends Relation<Inventory, WorkEvent>
{
    sort: number;
    constructor(ownerId: string, targetId: string) { 
        super();
        this.ownerId = ownerId;
        this.targetId = targetId;
    }
} 